File: sdks/php/examples/basic_usage.php
<?php

declare(strict_types=1);

require __DIR__ . '/../vendor/autoload.php';

use Nexus\SDK\Config;
use Nexus\SDK\NexusClient;
use Nexus\SDK\NexusApiException;

// Create client
$config = new Config(
    baseUrl: 'http://localhost:15474',
    apiKey: 'demo-api-key', // Replace with your API key
    timeout: 30
);

$client = new NexusClient($config);

try {
    // Check connection
    echo "Connecting to Nexus...\n";
    $client->ping();
    echo "✓ Connected successfully\n\n";

    // Create nodes — createNode now returns CreateNodeResponse {nodeId, message, error}.
    echo "--- Creating Nodes ---\n";
    $alice = $client->createNode(
        labels: ['Person'],
        properties: [
            'name' => 'Alice',
            'age' => 28,
            'city' => 'San Francisco'
        ]
    );
    echo "Created: Alice (ID: {$alice->nodeId})\n";

    $bob = $client->createNode(
        labels: ['Person'],
        properties: [
            'name' => 'Bob',
            'age' => 32,
            'city' => 'New York'
        ]
    );
    echo "Created: Bob (ID: {$bob->nodeId})\n";

    // Create relationship
    echo "\n--- Creating Relationship ---\n";
    $rel = $client->createRelationship(
        startNode: $alice->nodeId,
        endNode: $bob->nodeId,
        type: 'KNOWS',
        properties: [
            'since' => '2020',
            'strength' => 0.8
        ]
    );
    echo "Created: Alice -[{$rel->type}]-> Bob\n";

    // Query data
    echo "\n--- Querying Data ---\n";
    $result = $client->executeCypher(
        query: 'MATCH (p:Person) WHERE p.age > $minAge RETURN p.name as name, p.age as age, p.city as city ORDER BY p.age',
        parameters: ['minAge' => 25]
    );

    echo "Found " . count($result->rows) . " people older than 25:\n";
    foreach ($result->rows as $row) {
        echo "  - {$row['name']}, {$row['age']} years old, from {$row['city']}\n";
    }
    echo "Query took {$result->stats?->executionTimeMs}ms\n";

    // Get node by ID — returns GetNodeResponse {node, message, error}.
    echo "\n--- Reading Node ---\n";
    $resp = $client->getNode($alice->nodeId);
    if ($resp->node !== null) {
        echo "Retrieved: {$resp->node->properties['name']}\n";
    }

    // Update node — returns the raw {message, error} envelope.
    echo "\n--- Updating Node ---\n";
    $client->updateNode(
        id: $alice->nodeId,
        properties: [
            'age' => 29,
            'city' => 'Los Angeles'
        ]
    );
    echo "Updated Alice's age to 29 (Los Angeles)\n";

    // Query with relationships
    echo "\n--- Querying Relationships ---\n";
    $result = $client->executeCypher(
        'MATCH (a:Person)-[r:KNOWS]->(b:Person) RETURN a.name as person1, r.since as since, b.name as person2'
    );

    echo "Found " . count($result->rows) . " relationships:\n";
    foreach ($result->rows as $row) {
        echo "  {$row['person1']} knows {$row['person2']} since {$row['since']}\n";
    }

    // Cleanup
    echo "\n--- Cleanup ---\n";
    $client->deleteRelationship($rel->id);
    echo "✓ Deleted relationship\n";

    $client->deleteNode($alice->nodeId);
    echo "Deleted Alice\n";

    $client->deleteNode($bob->nodeId);
    echo "Deleted Bob\n";

    echo "\n✓ Example completed successfully\n";

} catch (NexusApiException $e) {
    echo "API Error: HTTP {$e->statusCode}: {$e->responseBody}\n";
} catch (Exception $e) {
    echo "Error: {$e->getMessage()}\n";
}

File: sdks/go/examples/basic_usage/basic_usage.go
package main

import (
	"context"
	"fmt"
	"log"
	"time"

	nexus "github.com/hivellm/nexus-go"
)

func main() {
	// Create client
	client := nexus.NewClient(nexus.Config{
		BaseURL: "http://localhost:15474",
		APIKey:  "demo-api-key", // Replace with your API key
		Timeout: 30 * time.Second,
	})

	ctx := context.Background()

	// Check connection
	fmt.Println("Connecting to Nexus...")
	if err := client.Ping(ctx); err != nil {
		log.Fatal("Failed to connect:", err)
	}
	fmt.Println("✓ Connected successfully")

	// Create nodes
	fmt.Println("\n--- Creating Nodes ---")
	alice, err := client.CreateNode(ctx, []string{"Person"}, map[string]interface{}{
		"name": "Alice",
		"age":  28,
		"city": "San Francisco",
	})
	if err != nil {
		log.Fatal(err)
	}
	fmt.Printf("Created: %s (ID: %s)\n", alice.Properties["name"], alice.ID)

	bob, err := client.CreateNode(ctx, []string{"Person"}, map[string]interface{}{
		"name": "Bob",
		"age":  32,
		"city": "New York",
	})
	if err != nil {
		log.Fatal(err)
	}
	fmt.Printf("Created: %s (ID: %s)\n", bob.Properties["name"], bob.ID)

	// Create relationship
	fmt.Println("\n--- Creating Relationship ---")
	rel, err := client.CreateRelationship(ctx, alice.ID, bob.ID, "KNOWS", map[string]interface{}{
		"since":    "2020",
		"strength": 0.8,
	})
	if err != nil {
		log.Fatal(err)
	}
	fmt.Printf("Created: %s -[%s]-> %s\n", alice.Properties["name"], rel.Type, bob.Properties["name"])

	// Query data
	fmt.Println("\n--- Querying Data ---")
	result, err := client.ExecuteCypher(ctx, `
		MATCH (p:Person)
		WHERE p.age > $minAge
		RETURN p.name as name, p.age as age, p.city as city
		ORDER BY p.age
	`, map[string]interface{}{
		"minAge": 25,
	})
	if err != nil {
		log.Fatal(err)
	}

	fmt.Printf("Found %d people older than 25:\n", len(result.Rows))
	for _, row := range result.RowsAsMap() {
		fmt.Printf("  - %s, %v years old, from %s\n", row["name"], row["age"], row["city"])
	}
	fmt.Printf("Query took %.2fms\n", result.Stats.ExecutionTimeMs)

	// Get node by ID
	fmt.Println("\n--- Reading Node ---")
	node, err := client.GetNode(ctx, alice.ID)
	if err != nil {
		log.Fatal(err)
	}
	fmt.Printf("Retrieved: %+v\n", node.Properties)

	// Update node
	fmt.Println("\n--- Updating Node ---")
	updated, err := client.UpdateNode(ctx, alice.ID, map[string]interface{}{
		"age":  29,
		"city": "Los Angeles",
	})
	if err != nil {
		log.Fatal(err)
	}
	fmt.Printf("Updated: %s is now %v years old and lives in %s\n",
		updated.Properties["name"], updated.Properties["age"], updated.Properties["city"])

	// Query with relationships
	fmt.Println("\n--- Querying Relationships ---")
	result, err = client.ExecuteCypher(ctx, `
		MATCH (a:Person)-[r:KNOWS]->(b:Person)
		RETURN a.name as person1, r.since as since, b.name as person2
	`, nil)
	if err != nil {
		log.Fatal(err)
	}

	fmt.Printf("Found %d relationships:\n", len(result.Rows))
	for _, row := range result.RowsAsMap() {
		fmt.Printf("  %s knows %s since %s\n", row["person1"], row["person2"], row["since"])
	}

	// Cleanup
	fmt.Println("\n--- Cleanup ---")
	if err := client.DeleteRelationship(ctx, rel.ID); err != nil {
		log.Fatal(err)
	}
	fmt.Println("✓ Deleted relationship")

	if err := client.DeleteNode(ctx, alice.ID); err != nil {
		log.Fatal(err)
	}
	fmt.Println("✓ Deleted Alice")

	if err := client.DeleteNode(ctx, bob.ID); err != nil {
		log.Fatal(err)
	}
	fmt.Println("✓ Deleted Bob")

	fmt.Println("\n✓ Example completed successfully")
}
